src/cpp/statistics/fastdds/domain/DomainParticipant.cpp

```diff
@@ -78,83 +79,6 @@ const DomainParticipant* DomainParticipant::narrow(
 #endif // FASTDDS_STATISTICS
 }
 
-ReturnCode_t DomainParticipant::enable()
-{
-    ReturnCode_t ret = eprosima::fastdds::dds::DomainParticipant::enable();
-
-    if (enable_)
-    {
-        create_statistics_builtin_entities();
-    }
-
-    return ret;
-}
-
-void DomainParticipant::create_statistics_builtin_entities()
-{
-    // Builtin publisher
-    builtin_publisher_ = impl_->create_publisher(eprosima::fastdds::dds::PUBLISHER_QOS_DEFAULT);
-
-    // Enable statistics datawriters
-    // 1. Find fastdds_statistics PropertyPolicyQos
-    const std::string* property_topic_list = eprosima::fastrtps::rtps::PropertyPolicyHelper::find_property(
-        impl_->get_qos().properties(), "fastdds.statistics");
-
-    if (nullptr != property_topic_list)
-    {
-        enable_statistics_builtin_datawriters(*property_topic_list);
-    }
-
-    // 2. FASTDDS_STATISTICS environment variable
-    std::string env_topic_list;
-    const char* data;
-    if (ReturnCode_t::RETCODE_OK == SystemInfo::get_env(FASTDDS_STATISTICS_ENVIRONMENT_VARIABLE, &data))
-    {
-        env_topic_list = data;
-    }
-
-    if (!env_topic_list.empty())
-    {
-        enable_statistics_builtin_datawriters(env_topic_list);
-    }
-}
-
-void DomainParticipant::enable_statistics_builtin_datawriters(
-        const std::string& topic_list)
-{
-    // Parse list and call enable_statistics_datawriter
-    std::stringstream topics(topic_list);
-    std::string topic;
-    while (std::getline(topics, topic, ';'))
-    {
-        ReturnCode_t ret = enable_statistics_datawriter(topic, STATISTICS_DATAWRITER_QOS);
-        // case RETCODE_ERROR is checked and logged in enable_statistics_datawriter.
-        // case RETCODE_INCONSISTENT_POLICY cannot happen. STATISTICS_DATAWRITER_QOS is consitent.
-        // case RETCODE_UNSUPPORTED cannot happen because this method is only called if FASTDDS_STATISTICS
-        // CMake option is enabled
-        assert(ret != ReturnCode_t::RETCODE_INCONSISTENT_POLICY);
-        assert(ret != ReturnCode_t::RETCODE_UNSUPPORTED);
-        if (ret == ReturnCode_t::RETCODE_BAD_PARAMETER)
-        {
-            logError(STATISTICS_DOMAIN_PARTICIPANT, "Topic " << topic << "is not a valid statistics topic name/alias");
-        }
-    }
-}
-
-void DomainParticipant::delete_statistics_builtin_entities()
-{
-    std::vector<eprosima::fastdds::dds::DataWriter*> builtin_writers;
-    builtin_publisher_->get_datawriters(builtin_writers);
-    for (auto writer : builtin_writers)
-    {
-        std::string topic_name = writer->get_topic()->get_name();
```

Comment:
The usual practice is to delegate this sort of call to the implementation class where a mutex can prevent concurrency issues.
In this case, several threads running this method may invalidate the `writer` contents and lead to a segmentation fault.

Reply:
This code was already moved to the implementation class.